### PR TITLE
Re-write permissions if running as root

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following commit attempts to detect if you're executing with
sudo and to assign the lock file with the correct permissions, to
prevent cryptic error messages.

The code detects sudo using environment variables and correctly
parses the UID and GID from those values, before chown'ing the file.

If at any point there is an error whilst changing the owner, it
will continue as root and on the next run, it will give information
about the file in question. This should then aid diagnosing the
problem without changing the underlying existing flow.

CC: @howbazaar
Background information: https://bugs.launchpad.net/juju/+bug/1758369

### CODE REVIEW

File: mutex_flock.go [hunk 1]
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
...
+	}
+	open := func() (fd int, err error) {
+		fd, err = syscall.Open(flockName, syscall.O_CREAT|syscall.O_RDONLY|syscall.O_CLOEXEC, 0600)
+		if err != nil {
+			if os.IsPermission(err) {
+				err = errors.Annotatef(err, "unable to open %s", flockName)
+			}
+			return
+		}
+		// Attemptting to open a lock file as root whilst using sudo can cause
```
Comment (hmlanigan) [inline mutex_flock.go]: nit: attempting

File: mutex_flock.go [hunk 2]
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
...
+		// the lock file to have the wrong permissions. Subsequent calls to
```
Comment (hmlanigan) [inline mutex_flock.go]: the wrong permissions for a non-sudo user....

**Review by hmlanigan — COMMENTED** [review]
just a couple of nits, i realized i don't know enough about how sudo and the SUDO_* env vars work to know if this is a good approach.  :-/

File: mutex_flock.go [hunk 3]
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
 	}
 
 	flockName := filepath.Join(os.TempDir(), "juju-"+name)
+	chownFromRoot := func() (err error) {
```
Comment (howbazaar) [inline mutex_flock.go]: You don't need the named error response.

File: mutex_flock.go [hunk 4]
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
...
+		if cmd, ok := os.LookupEnv("SUDO_COMMAND"); ok && cmd != "" {
+			var uid, gid int
+			uid, err = strconv.Atoi(os.Getenv("SUDO_UID"))
+			if err != nil {
+				err = errors.Annotate(err, "parsing SUDO_UID")
```
Comment (howbazaar) [inline mutex_flock.go]: Even with named return values, I prefer the explicit approach.

File: mutex_flock.go [hunk 5]
```diff
@@ -67,8 +68,45 @@ func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
...
+			gid, err = strconv.Atoi(os.Getenv("SUDO_GID"))
+			if err != nil {
+				err = errors.Annotate(err, "parsing SUDO_GID")
+				return
+			}
+			return syscall.Chown(flockName, uid, gid)
+		}
+		return nil
+	}
+	open := func() (fd int, err error) {
```
Comment (howbazaar) [inline mutex_flock.go]: Can we not use named return values and just have explicit returns? There are only two.

File: mutex_test.go [hunk 6]
```diff
@@ -284,6 +287,27 @@ func (s *mutexSuite) TestMutexNotInherited(c *gc.C) {
 	r.Release()
 }
 
+// TestFilePermissions tests that the file permissions are correct.
+func (s *mutexSuite) TestFilePermissions(c *gc.C) {
```
Comment (howbazaar) [inline mutex_test.go:294]: While kinda useful, this test doesn't actually test the code you have written at all because it will never be run with sudo.

**Review by howbazaar — APPROVED** [review]
I think the approach is fine, but I'd like the functions not to use named return values.

File: mutex_test.go (hunk 6 above)
Comment (SimonRichardson) [reply mutex_test.go:294]: Yeah, I'll make a complimentary test function, good spot.

File: mutex_flock.go (hunk 5 above)
Comment (SimonRichardson) [reply mutex_flock.go]: Of course

File: mutex_flock.go (hunk 4 above)
Comment (SimonRichardson) [reply mutex_flock.go]: Fine by me, I'll get on it!

File: mutex_test.go (hunk 6 above)
Comment (SimonRichardson) [reply mutex_test.go:294]: fixed in https://github.com/juju/mutex/pull/6/commits/b5c51c193415eabe150a62058ec2985c57395433

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**